proctor-common/src/main/java/com/indeed/proctor/common/Proctor.java

```diff
@@ -325,34 +329,45 @@ public ProctorResult determineTestGroups(
         }
 
         final Map<String, ValueExpression> localContext =
-                ProctorUtils.convertToValueExpressionMap(
+                ProctorUtils.convertLocalContextToValueExpressionMap(
                         RuleEvaluator.EXPRESSION_FACTORY, inputContext);
 
         for (final String testName : filteredEvaluationOrder) {
             final TestChooser<?> testChooser = testChoosers.get(testName);
             final String identifier;
+            final boolean containsUnitlessAllocations =
+                    testTypesWithInvalidIdentifier.contains(
+                                    testChooser.getTestDefinition().getTestType())
+                            && testChooser.getTestDefinition().getContainsUnitlessAllocation();
+            if (!containsUnitlessAllocations) {
+                localContext.put(
```

Comment:
whether to have containsUnitlessAllocations is based on test types. Once the UNITLESS_ALLOCATION_IDENTIFIER is set to true for a given test, does it need to be reset for next test?

Reply:
It is based on test type of the test definition and if the test definition has enabledUnitlessAllocations and an allocation with a rule containing missingExperimentalUnit. 

It does not need to be reset as it is already unique for each test.

Reply:
"test type of the test definition and if the test definition has enabledUnitlessAllocations and an allocation with a rule containing missingExperimentalUnit"

for test A that does not have missingExperimentalUnit in the rule, now we set
localContext.put(
                        UNITLESS_ALLOCATION_IDENTIFIER,
                        RuleEvaluator.EXPRESSION_FACTORY.createValueExpression(
                                false, Object.class));


For next test, test B, which has missingExperimentalUnit in the rule, do we set it back to true?
localContext.put(
                        UNITLESS_ALLOCATION_IDENTIFIER,
                        RuleEvaluator.EXPRESSION_FACTORY.createValueExpression(
                                true, Object.class));

Reply:
Ah I see misunderstood the initial comment was focusing on the boolean variable rather than setting the context. Good call out. Updated logic to always set if it not equal. Tried to make it efficient as this is logic hit very often by making static variables for the ValueExpression to avoid reconstructing every time it needs to switch.

Reply:
"by making static variables for the ValueExpression to avoid reconstructing every time it needs to switch"

Thanks that's a great optimization.